python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1762,6 +1763,105 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
             )
         return primary
 
+    def getManyURIs(
+        self,
+        refs: Iterable[DatasetRef],
+        predict: bool = False,
+        throwIfNotFound: bool = True,
+    ) -> Dict[DatasetRef, DatasetRefURIs]:
+        # Docstring inherited
+
+        PrimaryType = Optional[ResourcePath]
+        ComponentsType = Dict[str, ResourcePath]
+        uris: Dict[DatasetRef, DatasetRefURIs] = {}
+
+        records = self._get_stored_records_associated_with_refs(refs)
+
+        missingRefs = [ref for ref in refs if ref.id not in records.keys()]
+
+        for ref in missingRefs:
+            # if this has never been written then we have to guess
+            if not self.exists(ref) and not predict:
+                if throwIfNotFound:
+                    raise FileNotFoundError("Dataset {} not in this datastore.".format(ref))
+                else:
+                    continue
+
+            primary: PrimaryType = None
+            components: ComponentsType = {}
+
+            if self.composites.shouldBeDisassembled(ref):
+
+                for component, _ in ref.datasetType.storageClass.components.items():
+                    compRef = ref.makeComponentRef(component)
+                    compLocation, _ = self._determine_put_formatter_location(compRef)
+
+                    # Add the "#predicted" URI fragment to indicate this is a
+                    # guess
+                    components[component] = ResourcePath(compLocation.uri.geturl() + "#predicted")
+
+            else:
+
+                try:
+                    location, _ = self._determine_put_formatter_location(ref)
+                except DatasetTypeNotSupportedError:
+                    if throwIfNotFound:
+                        raise
+                    else:
+                        continue
+
+                # Add the "#predicted" URI fragment to indicate this is a guess
+                primary = ResourcePath(location.uri.geturl() + "#predicted")
+
+            uris[ref] = DatasetRefURIs(primary, components)
+
+        for ref in refs:
+            if ref.id is None:
+                log.warning("Can't get information about a ref when ref.id is None: %s", ref)
+                continue
+            storedFileInfos = records.get(ref.id, None)
+            if storedFileInfos is None:
+                continue
+
+            fileLocations = [(i.file_location(self.locationFactory), i) for i in storedFileInfos]
+
+            guessing = False
```

Comment:
Isn't this all duplicating the logic from the second half of getURIs? We can't be duplicating all this. Can you factor getURIs into a prediction part and "must exist" part so they can both share the same logic? The first new method takes a DatasetRef that is known not to exist, the second can take a StoredFileInfo.